iml-agent/src/action_plugins/check_ha.rs

```diff
@@ -71,27 +71,146 @@ impl AgentInfo {
     }
 }
 
+fn _check_ha(output: &[u8]) -> Result<Vec<AgentInfo>, ImlAgentError> {
+    match Element::from_reader(output) {
+        Err(err) => Err(ImlAgentError::XmlError(err)),
+        Ok(elem) => Ok(elem
+            .find_all("group")
+            .flat_map(|g| {
+                let name = g.get_attr("id").unwrap_or("").to_string();
+                g.find_all("primitive")
+                    .map(move |p| AgentInfo::create(p, Some(name.clone())))
+            })
+            .chain(
+                elem.find_all("primitive")
+                    .map(|p| AgentInfo::create(p, None)),
+            )
+            .collect()),
+    }
+}
+
 pub fn check_ha(_: ()) -> impl Future<Item = Vec<AgentInfo>, Error = ImlAgentError> {
     cmd_output_success("cibadmin", &["--query", "--xpath", "//resources"])
         .instrument(span!(Level::INFO, "Read cib"))
-        .and_then(|output| {
-            // This cannot split between map/map_err because Element does not implement Send
-            match Element::from_reader(output.stdout.as_slice()) {
-                Err(err) => future::err(ImlAgentError::XmlError(err)),
-                Ok(elem) => future::ok(
-                    elem.find_all("group")
-                        .flat_map(|g| {
-                            let name = g.get_attr("id").unwrap_or("").to_string();
-                            g.find_all("primitive")
-                                .map(move |p| AgentInfo::create(p, Some(name.clone())))
-                        })
-                        .chain(
-                            elem.find_all("primitive")
-                                .map(|p| AgentInfo::create(p, None)),
-                        )
-                        .collect(),
+        .and_then(|output| _check_ha(&output.stdout.as_slice()))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{AgentInfo, ResourceAgentType, _check_ha};
+    use crate::agent_error;
+    use std::collections::HashMap;
+
+    #[test]
+    fn test_ha_only_fence_chroma() -> agent_error::Result<()> {
```

Comment:
Any reason you need to return an agent_error::Result? I think you could just do:

```
fn test_ha_only_fence_chroma() {
        let testxml = r#"<resources>
  <primitive class="stonith" id="st-fencing" type="fence_chroma"/>
</resources>
"#;
        assert_eq!(
            _check_ha(&testxml.as_bytes()).unwrap(),
            vec![AgentInfo {
                agent: ResourceAgentType::new(
                    "stonith".to_string(),
                    None,
                    "fence_chroma".to_string()
                ),
                group: None,
                id: "st-fencing".to_string(),
                args: HashMap::new()
            }]
        );
}
```

Reply:
Yeah, since this test it's not using `?` it doesn't need to return a result.